src/template_finder.py

```diff
@@ -89,6 +96,104 @@ def search(
+        :param timeout: After this amount of time the search will stop and it will return [False, None]
+        :param take_ss: Bool value to take screenshot on timeout or not (flag must still be set in params!)
+        Other params are the same as for TemplateFinder.search()
+        """
+        if type(ref) is str:
+            ref = [ref]
+        if not suppress_debug:
+            Logger.debug(f"Waiting for templates: {ref}")
+        start = time.time()
+        while 1:
+            img = grab()
+            template_match = self.search(ref, img, roi=roi, threshold=threshold, best_match=best_match, use_grayscale=use_grayscale, normalize_monitor=normalize_monitor, color_match=color_match)
+            is_loading_black_roi = np.average(img[:, 0:Config().ui_roi["loading_left_black"][2]]) < 1.0
+            if not is_loading_black_roi or "LOADING" in ref:
+                if template_match.valid:
+                    Logger.debug(f"Found Match: {template_match.name} ({template_match.score*100:.1f}% confidence)")
+                    return template_match
+                if timeout is not None and (time.time() - start) > timeout:
+                    if Config().general["info_screenshots"] and take_ss:
+                        cv2.imwrite(f"./info_screenshots/info_wait_for_{ref}_timeout_" + time.strftime("%Y%m%d_%H%M%S") + ".png", img)
+                    if take_ss:
+                        Logger.debug(f"Could not find any of the above templates")
+                    return template_match
+
+    def search_all(
+            self,
+            ref: Union[str, np.ndarray, list[str]],
+            inp_img: np.ndarray,
+            threshold: float = 0.68,
+            roi: list[float] = None,
+            normalize_monitor: bool = False,
+            use_grayscale: bool = False,
+            color_match: list = False
+    ) -> list[TemplateMatch]:
+        """
+        Search for a template in an image
+        :param ref: Either key of a already loaded template, list of such keys, or a image which is used as template
+        :param inp_img: Image in which the template will be searched
+        :param threshold: Threshold which determines if a template is found or not
+        :param roi: Region of Interest of the inp_img to restrict search area. Format [left, top, width, height]
+        :param normalize_monitor: If True will return positions in monitor coordinates. Otherwise in coordinates of the input image.
+        :param use_grayscale: Use grayscale template matching for speed up
+        :param color_match: Pass a color to be used by misc.color_filter to filter both image of interest and template image (format Config().colors["color"])
+        :return: Returns a list of TemplateMatch objects with a valid flag
+        """
+        return self._search_all_internal(ref, inp_img, threshold, roi, normalize_monitor, _SearchType.ALL_MATCH, use_grayscale, color_match)
```

Comment:
Whats to point of having an private (internal) function that is just mapped from a public one?

Reply:
1. Not duplicating code
2. Cleaner API by having:
a. a public seach by first / best match function that retuns a TemplateMatch 
b. another search_all function that returns a list of TemplateMatch